### ISSUE TITLE
Trial's intermediate_values are overwritten by other new trial when running multiple experiments in parallel

### ISSUE THREAD

**Issue by dktemp888** [optuna/optuna#1508]
Trial's intermediate_values are overwritten by other new trial when running multiple experiments in parallel
## Expected behavior
Each trial has its own intermediate_values that can be used for pruner correctly when running Optuna in parallel. This is important because otherwise the pruner logic will be wrong because of incorrect intermediate_values.

## Environment
Ubuntu 1604

- Optuna version: 1.14, 1.15
- Python version: 3.6, 3.7
- OS: Ubuntu 1604
- (Optional) Other libraries and their versions:

## Error messages, stack traces, or logs

```
# error messages, stack traces, or logs
```

## Steps to reproduce
Assume you have a training code with Optuna as hyper-parameter optimizer
1. run the first instance of train code with optuna;
2. run the second instance of train code with optuna;
3. observe the trial.intermediate_values dict, the first instance's intermediate_values are overwritten by the second. However if you check the database file all intermediate values and steps are correctly logged in the database. Only the runtime value is wrong.

It happens on both sqlite and mysql.

## Reproducible examples (optional)

```python
# python code
```

**Comment by hvy** [comment]
Thanks for the report and for verifying with both SQLite and MySQL. At a quick glance, I'm wondering if it's not a different manifestation of #1488. If you are able to, could you try to install the latest `master` branch (check out the `master` branch and run `pip install -e .`) of Optuna where #1488 is fixed and check if the bug persists?

In case this is a different bug, it's a quite severe one. Are you able to give some more details on the reproduction, preferably in code? For instance, I'd be good to know how you run your experiments in parallel, in which runtime object, belonging to which worker, you're observing the different intermediate values.

**Comment by dktemp888** [comment]
Hi @hvy thanks so much for your immediate response. Based on your suggestion I ran with the latest master. It still has the issue. Unfortunately I cannot share my current version source code for the moment. Below is the log that can help show some level of details.

Regarding how to run experiments in parallel: 
(1) this issue was discovered in a distributed training cluster using Optuna, however I can reproduce it even in a single machine running multiple instances;
(2) say we have train.py with optuna as hyper parameter optimizer, and I have my own pruner algorithm implemented which subclass from optuna.pruners.BasePruner, and uses trial.intermediate_values as input to my pruner;
(3) within my own pruner class's prune(self, study, trial) method, I print the intermediate_values from its trial.intermediate_values --> that's what below log shows;
(4) below log was captured just now by running the same train.py in two consoles. Since they share the same database, that's the recommended way to do parallel hyper-parameter optimization from Optuna.



============================== BELOW IS FROM CONSOLE #1, WHICH RUNS FIRST ===========================================

2020-07-13 10:09:35,025 [INFO] resume from previous run containing 14 trials...
2020-07-13 10:09:35,057 [INFO] No.14 run, cmd: python iris.py 20200713-100934-iris-7000-14000-48000.txt 0.0026805064435061767 7000 14000 48000, log: 20200713-100934-iris-7000-14000-48000.txt
2020-07-13 10:09:35,057 [INFO] No.14 run, submitting job...

2020-07-13 10:09:45,087 [INFO] No.14 run, update from job, step = 0, loss = 0.421053 ... push into queue.
2020-07-13 10:09:45,088 [INFO] No.14 run, received update from job: step = 0, loss = 0.421053.
2020-07-13 10:09:45,236 [INFO] trial intermediate value: {0: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:09:50,092 [INFO] No.14 run, update from job, step = 1, loss = 0.421053 ... push into queue.
2020-07-13 10:09:50,092 [INFO] No.14 run, received update from job: step = 1, loss = 0.421053.
2020-07-13 10:09:50,230 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:09:55,097 [INFO] No.14 run, update from job, step = 2, loss = 0.078947 ... push into queue.
2020-07-13 10:09:55,098 [INFO] No.14 run, received update from job: step = 2, loss = 0.078947.
2020-07-13 10:09:55,229 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.078947}, least trial in top runs: 0.2105.

2020-07-13 10:10:00,102 [INFO] No.14 run, update from job, step = 3, loss = 0.421053 ... push into queue.
2020-07-13 10:10:00,102 [INFO] No.14 run, received update from job: step = 3, loss = 0.421053.
2020-07-13 10:10:00,254 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.078947, 3: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:10:05,103 [INFO] No.14 run, update from job, step = 4, loss = 0.421053 ... push into queue.
2020-07-13 10:10:05,104 [INFO] No.14 run, received update from job: step = 4, loss = 0.421053.
2020-07-13 10:10:05,256 [INFO] trial intermediate value: {}, least trial in top runs: 0.2105.

2020-07-13 10:10:10,110 [INFO] No.14 run, update from job, step = 5, loss = 0.289474 ... push into queue.
2020-07-13 10:10:10,110 [INFO] No.14 run, received update from job: step = 5, loss = 0.289474.
2020-07-13 10:10:10,257 [INFO] trial intermediate value: {}, least trial in top runs: 0.2105.

2020-07-13 10:10:15,114 [INFO] No.14 run, update from job, step = 6, loss = 0.078947 ... push into queue.
2020-07-13 10:10:15,115 [INFO] No.14 run, received update from job: step = 6, loss = 0.078947.
2020-07-13 10:10:15,271 [INFO] trial intermediate value: {0: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:10:20,120 [INFO] No.14 run, update from job, step = 7, loss = 0.421053 ... push into queue.
2020-07-13 10:10:20,120 [INFO] No.14 run, received update from job: step = 7, loss = 0.421053.
2020-07-13 10:10:20,275 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:10:25,126 [INFO] No.14 run, update from job, step = 8, loss = 0.421053 ... push into queue.
2020-07-13 10:10:25,127 [INFO] No.14 run, received update from job: step = 8, loss = 0.421053.
2020-07-13 10:10:25,275 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.473684}, least trial in top runs: 0.2105.

2020-07-13 10:10:30,131 [INFO] No.14 run, update from job, step = 9, loss = 0.421053 ... push into queue.
2020-07-13 10:10:30,131 [INFO] No.14 run, received update from job: step = 9, loss = 0.421053.
2020-07-13 10:10:30,254 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.473684, 3: 0.289474}, least trial in top runs: 0.2105.

2020-07-13 10:10:35,136 [INFO] No.14 run, update from job, step = 10, loss = 0.052632 ... push into queue.
2020-07-13 10:10:35,136 [INFO] No.14 run, received update from job: step = 10, loss = 0.052632.
2020-07-13 10:10:35,292 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.473684, 3: 0.289474, 4: 0.421053}, least trial in top runs: 0.2105.




============================== BELOW IS FROM CONSOLE #2, WHICH RUNS SLIGHTLY LATER ===================================

2020-07-13 10:10:02,487 [INFO] resume from previous run containing 15 trials...
2020-07-13 10:10:02,528 [INFO] No.15 run, cmd: python iris.py 20200713-101002-iris-5000-28000-56000.txt 1.8277337702484806e-05 5000 28000 56000, log: 20200713-101002-iris-5000-28000-56000.txt
2020-07-13 10:10:02,528 [INFO] No.15 run, submitting job...

2020-07-13 10:10:12,560 [INFO] No.15 run, update from job, step = 0, loss = 0.421053 ... push into queue.
2020-07-13 10:10:12,560 [INFO] No.15 run, received update from job: step = 0, loss = 0.421053.
2020-07-13 10:10:12,685 [INFO] trial intermediate value: {0: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:10:17,568 [INFO] No.15 run, update from job, step = 1, loss = 0.421053 ... push into queue.
2020-07-13 10:10:17,569 [INFO] No.15 run, received update from job: step = 1, loss = 0.421053.
2020-07-13 10:10:17,692 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:10:22,575 [INFO] No.15 run, update from job, step = 2, loss = 0.473684 ... push into queue.
2020-07-13 10:10:22,575 [INFO] No.15 run, received update from job: step = 2, loss = 0.473684.
2020-07-13 10:10:22,698 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.473684}, least trial in top runs: 0.2105.

2020-07-13 10:10:27,577 [INFO] No.15 run, update from job, step = 3, loss = 0.289474 ... push into queue.
2020-07-13 10:10:27,577 [INFO] No.15 run, received update from job: step = 3, loss = 0.289474.
2020-07-13 10:10:27,701 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.473684, 3: 0.289474}, least trial in top runs: 0.2105.

2020-07-13 10:10:32,581 [INFO] No.15 run, update from job, step = 4, loss = 0.421053 ... push into queue.
2020-07-13 10:10:32,582 [INFO] No.15 run, received update from job: step = 4, loss = 0.421053.
2020-07-13 10:10:32,715 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.473684, 3: 0.289474, 4: 0.421053}, least trial in top runs: 0.2105.

2020-07-13 10:10:37,586 [INFO] No.15 run, update from job, step = 5, loss = 0.394737 ... push into queue.
2020-07-13 10:10:37,587 [INFO] No.15 run, received update from job: step = 5, loss = 0.394737.
2020-07-13 10:10:37,712 [INFO] trial intermediate value: {0: 0.421053, 1: 0.421053, 2: 0.473684, 3: 0.289474, 4: 0.421053, 5: 0.394737}, least trial in top runs: 0.2105.

**Comment by hvy** [comment]
Thanks for your quick response. Sad new though, I'll have a look at it today.

**Comment by hvy** [comment]
So it seems like the intermediate values in the first process is cleared when the second process starts? 

Sorry but I did not manage to reproduce the same behavior.

Given that the intermediate values are correctly written to the database, I don't really see how your second Python process manages to alter the intermediate values seen by the first process. If I've understood it correctly that is. Do they happen to share memory? Also, mostly wondering but why does the above log only includes the intermediate values from a single trial? If you have multiple processes, they should in general be assigned different trials, i.e. a single trial should in general only be managed by a single process/thread. Maybe you just filtered the log?

**Comment by dktemp888** [comment]
I forgot to mention that all the intermediate values in database are still correctly captured. The issue only happens within the running instance where you get it from trial.intermediate_values property.

I ran two instances separately and the console logs shown above were captured from individual instance. There are separators: "BELOW IS FROM CONSOLE #1, WHICH RUNS FIRST", "BELOW IS FROM CONSOLE #2, WHICH RUNS SLIGHTLY LATER".
There is no memory sharing between two processes. and you're right each process is assigned with different trials.

As the next step, is it possible for you try printing the trial.intermediate_values from memory instead of checking from database?

If not I will need to extract a minimum code and share to you.

**Comment by hvy** [comment]
> As the next step, is it possible for you try printing the trial.intermediate_values from memory instead of checking from database?

Sorry, I don't quite understand what you mean here, but as you say, minimum reproduction code would be really helpful. Would you possibly be able to extract that?

**Comment by dktemp888** [comment]
During the extraction of reproducible code, I discovered a bug in our code. The trail variable was overwritten during a dict sort within the pruner.prune() function. That's why the intermediate_values are 'overwritten' later on...

I am really sorry for the confusion and thanks again for your prompt support.

This issue can be closed now.

**Comment by hvy** [comment]
That's a relief. No worries and thanks for the update!
